ksql-streams/src/main/java/io/confluent/ksql/execution/streams/SinkBuilder.java

```diff
@@ -0,0 +1,130 @@
+import org.apache.kafka.streams.kstream.Produced;
+import org.apache.kafka.streams.kstream.Transformer;
+import org.apache.kafka.streams.kstream.TransformerSupplier;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.To;
+
+public final class SinkBuilder {
+  private SinkBuilder() {
+  }
+
+  public static  <K> void build(
+      final LogicalSchema schema,
+      final Formats formats,
+      final Optional<TimestampColumn> timestampColumn,
+      final String topicName,
+      final KStream<K, GenericRow> stream,
+      final KeySerdeFactory<K> keySerdeFactory,
+      final QueryContext queryContext,
+      final KsqlQueryBuilder queryBuilder
+  ) {
+    final PhysicalSchema physicalSchema = PhysicalSchema.from(schema, formats.getOptions());
+
+    final Serde<K> keySerde = keySerdeFactory.buildKeySerde(
+        formats.getKeyFormat(),
+        physicalSchema,
+        queryContext
+    );
+
+    final Serde<GenericRow> valueSerde = queryBuilder.buildValueSerde(
+        formats.getValueFormat(),
+        physicalSchema,
+        queryContext
+    );
+
+    final int timestampColumnIndex = timestampColumn.map(TimestampColumn::getColumn)
+        .map(c -> schema.findValueColumn(c).orElseThrow(IllegalStateException::new))
+        .map(Column::index)
+        .orElse(-1);
+
+    stream.transform(new TransformTimestamp<>(timestampColumnIndex))
+        .to(topicName, Produced.with(keySerde, valueSerde));
+  }
+
+  static class TransformTimestamp<K>
+      implements TransformerSupplier<K, GenericRow, KeyValue<K, GenericRow>> {
+    private final int timestampColumnIndex;
+
+    TransformTimestamp(final int timestampColumnIndex) {
+      this.timestampColumnIndex = timestampColumnIndex;
+    }
+
+    @Override
+    public boolean equals(final Object o) {
+      if (o == null || !(o instanceof TransformTimestamp)) {
+        return false;
+      }
+
+      final TransformTimestamp that = (TransformTimestamp)o;
+      return timestampColumnIndex == that.timestampColumnIndex;
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hashCode(timestampColumnIndex);
+    }
+
+    @Override
+    public Transformer<K, GenericRow, KeyValue<K, GenericRow>> get() {
+      return new Transformer<K, GenericRow, KeyValue<K, GenericRow>>() {
+        private ProcessorContext processorContext;
+
+        @Override
+        public void init(final ProcessorContext processorContext) {
+          this.processorContext = requireNonNull(processorContext, "processorContext");
+        }
+
+        @Override
+        public KeyValue<K, GenericRow> transform(final K key, final GenericRow row) {
+          if (timestampColumnIndex >= 0 && row.get(timestampColumnIndex) instanceof Long) {
+            processorContext.forward(
+                key,
+                row,
+                To.all().withTimestamp((long) row.get(timestampColumnIndex))
+            );
```

Comment:
I think the logic here should write the `System.currenttimeMillis()` in case there is no timestamp column set, right? Otherwise, the output ROWTIME is derived from the source topics.

Reply:
I don't think so. From my understanding the `WITH ROWTIME` clause in `CTAS/CSAS` statement _overwrite_ the existing event-timestamp with something different. Hence, if no `WITH ROWTIME` column is specified the input key and value should be forwarded without a modified timestamp?

Reply:
Is this an out of date comment? It doesn't look to match the current impl...